File: DSA/DSA.DataStructures/Linear/Queue/LLQueue/Queue.cs
﻿using DSA.DataStructures.Linear.LinkedLists.DoubleLinkedListNS;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DSA.DataStructures.Linear.Queue.LLQueue
{
    public class Queue : IQueue
    {
        private readonly DoubleLinkedList linkedList = new();
        public void Clear()
        {
            linkedList.Clear();
        }

        public int Dequeue()
        {
            return linkedList.RemoveFirst();
        }

        public void Enqueue(int item)
        {
            linkedList.AddLast(item);
        }

        public bool IsEmpty()
        {
            return linkedList.GetLegnth() == 0;
        }

        public int Peek()
        {
            return linkedList.PeekFirst();
        }
    }
}

File: study/src/internal/datastructures/queue/queue.go
package queue

import "learningDSA/src/internal/datastructures/linkedlists"

type Queue struct {
	list linkedlists.DoubleLinkedList
}

func (q *Queue) Enqueue(d int) {
	q.list.AddLast(d)
}

func (q *Queue) Dequeue() (int, bool) {
	return q.list.RemoveFirst()
}

func (q *Queue) GetLength() int {
	return q.list.Length
}

func (q *Queue) Clear() {
	q.list.Clear()
}

func (q *Queue) String() string {
	return q.list.String()
}